lib/pgBackRest/Storage/S3/Request.pm

```diff
@@ -0,0 +1,202 @@
+            {name => 'strVerb', trace => true},
+            {name => 'strUri', optional => true, default => '/', trace => true},
+            {name => 'hQuery', optional => true, trace => true},
+            {name => 'hHeader', optional => true, trace => true},
+            {name => 'rstrBody', optional => true, trace => true},
+            {name => 'strResponseType', optional => true, default => S3_RESPONSE_TYPE_NONE, trace => true},
+            {name => 'bIgnoreMissing', optional => true, default => false, trace => true},
+        );
+
+    # Get datetime to be used for auth requests
+    my $strDateTime = s3DateTime();
+
+    # Set content length and hash
+    $hHeader->{&S3_HEADER_CONTENT_SHA256} = defined($rstrBody) ? sha256_hex($$rstrBody) : PAYLOAD_DEFAULT_HASH;
+    $hHeader->{&S3_HEADER_CONTENT_LENGTH} = defined($rstrBody) ? length($$rstrBody) : 0;
+
+    # Generate authorization header
+    $hHeader = s3AuthorizationHeader(
+        $self->{strRegion}, "$self->{strBucket}.$self->{strEndPoint}", $strVerb, $strUri, httpQuery($hQuery), $strDateTime,
+        $hHeader, $self->{strAccessKeyId}, $self->{strSecretAccessKey}, $hHeader->{&S3_HEADER_CONTENT_SHA256});
+
+    # Send the request
+    my $oHttpClient = new pgBackRest::Common::Http::Client(
+        $self->{strHost}, $strVerb,
+        {strUri => $strUri, hQuery => $hQuery, hRequestHeader => $hHeader, rstrRequestBody => $rstrBody,
+            bVerifySsl => $self->{bVerifySsl}, lBufferMax => $self->{lBufferMax}});
+
+    # Check response code
+    my $iReponseCode = $oHttpClient->responseCode();
+    my $oResponse;
+
+    if ($oHttpClient->responseCode() == 200)
+    {
+        # Save the response headers locally
+        $self->{hResponseHeader} = $oHttpClient->responseHeader();
+
+        # XML response is expected
+        if ($strResponseType eq S3_RESPONSE_TYPE_XML)
+        {
+            my $rtResponseBody = $oHttpClient->responseBody();
+
+            if ($oHttpClient->contentLength() == 0 || !defined($$rtResponseBody))
+            {
+                confess &log(ERROR,
+                    "response type '${strResponseType}' was requested but content length is zero or content is missing",
+                    ERROR_PROTOCOL);
+            }
+
+            $oResponse = xmlParse($$rtResponseBody);
+        }
+        # An IO object is expected for file responses
+        elsif ($strResponseType eq S3_RESPONSE_TYPE_IO)
+        {
+            $oResponse = $oHttpClient;
+        }
+    }
+    else
+    {
+        if ($oHttpClient->responseCode() == 404)
+        {
+            if (!$bIgnoreMissing)
+            {
+                confess &log(ERROR, "unable to open '${strUri}': No such file or directory", ERROR_FILE_MISSING);
+            }
+        }
+        else
+        {
+            my $rstrResponseBody = $oHttpClient->responseBody();
+
+            confess &log(ERROR,
+                "S3 request error [$iReponseCode] " . $oHttpClient->responseCode() .
```

Comment:
Won't this just read "S3 request error [123] 123 some body text"? why do you need to duplicate the response code?